### PR TITLE
[security] Allow to config web server's cipher and protocols

### PR DESCRIPTION
**Pull request by Shoothzj**
### Motivation
For the security, people want to config the support `ciphers` and `protocols`

### Modifications
- Allow the broker and proxy config web server's `ciphers` and `protocols`
- By default, `ciphers` and `protocols` are null, same as before.

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 
  
- [x] `doc` 
  
  this PR contains config file change. IMHO, the doc will be auto generated.

### CODE REVIEW

**Review by zymap — COMMENTED** [review]
Could you please add a test to cover this?

**Comment by Shoothzj** [comment]
> Could you please add a test to cover this?

Yes, add tests to test ciphers and protocols on jetty server.

**Comment by Shoothzj** [comment]
@codelipenghui @lhotari @michaeljmarshall @merlimat @315157973 PTAL, thanks

**Comment by Shoothzj** [comment]
@codelipenghui @lhotari @michaeljmarshall @merlimat @315157973 @hangc0276 PTAL, thanks

**Comment by Shoothzj** [comment]
@codelipenghui @lhotari @michaeljmarshall @merlimat @315157973 @hangc0276 PTAL, thanks

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                    }
+                }, 600);
+        factory.setHostnameVerifier((s, sslSession) -> true);
+        ServerConnector connector = new ServerConnector(server, factory);
+        int unusedPort = getUnusedPort();
+        connector.setPort(unusedPort);
+        connectors.add(connector);
+        server.setConnectors(connectors.toArray(new ServerConnector[0]));
+        server.start();
+        Thread.sleep(5000);
```
Comment (eolivelli) [inline pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:107]: can we remove this and use Awaiatility ?

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
```
Comment (eolivelli) [inline pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:146]: and this

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            trustManagerFactory.init(trustStore);
+            sslContext.init(keyManagers, trustManagerFactory.getTrustManagers(), new SecureRandom());
+            return sslContext;
+        } catch (Exception e) {
+            log.error("load ssl context error ", e);
+            return null;
+        }
+    }
+
+    private static int getUnusedPort() throws Exception {
```
Comment (eolivelli) [inline pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:193]: we already have a utility method for this

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall it looks good to me
I left some comments about the tests

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java (hunk 3 above)
Comment (Shoothzj) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:193]: @eolivelli Could you please tell me, where the method exists? I have searched for `unusedPort` and `freePort` etc. :)

Comment (Shoothzj) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:193]: @eolivelli I fould `getFreePort` in `PulsarMultiListenersWithInternalListenerNameTest` and `PulsarMultiHostClientTest`.
But now we don't have module like `test-common` or `test-util`. So I suggest now we put this method here. And start a new PR for `test-util` WDYT?

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java (hunk 1 above)
Comment (Shoothzj) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java]: I verified the code, It looks wo don't need sleep, the server is ready when `server.start()` returns. Remove the sleep.

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java (hunk 3 above)
Comment (eolivelli) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:190]: you can pass 0 and Jetty will get an ephemeral port, then you can ask to Jetty the actual LocalPort

Comment (eolivelli) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:190]: no need to add this method

**Review by eolivelli — COMMENTED** [review]
LGTM

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java (hunk 3 above)
Comment (Shoothzj) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java]: fixed

**Comment by Shoothzj** [comment]
@codelipenghui @lhotari @michaeljmarshall @merlimat @315157973 @hangc0276 PTAL, thanks

**Comment by Shoothzj** [comment]
@eolivelli PTAL, again, thanks

**Review by codelipenghui — APPROVED** [review]

**Comment by Shoothzj** [comment]
@eolivelli PTAL, again, thanks

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                    {
+                        this.add("TLSv1.3");
+                    }
+                }, 600);
+        factory.setHostnameVerifier((s, sslSession) -> true);
+        ServerConnector connector = new ServerConnector(server, factory);
+        connector.setPort(0);
+        connectors.add(connector);
+        server.setConnectors(connectors.toArray(new ServerConnector[0]));
+        server.start();
```
Comment (eolivelli) [inline pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:102]: Please stop all the servers and close all the clients

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java (hunk 4 above)
Comment (Shoothzj) [reply pulsar-common/src/test/java/org/apache/pulsar/common/util/keystoretls/JettySslContextFactoryWithAutoRefreshTest.java:104]: fixed

**Comment by Shoothzj** [comment]
@eolivelli PTAL, again, thanks

**Comment by Shoothzj** [comment]
@eolivelli PTAL, again, thanks

**Review by eolivelli — APPROVED** [review]
Lgtm

**Comment by BewareMyPower** [comment]
`JettySslContextFactoryWithAutoRefreshTest` failed with JDK 8.

image

It looks like our tests are all based on JDK 11 currently so that this test failure was not detected. And I found some other PRs that rely on this PR.

/cc @codelipenghui @eolivelli @nodece @michaeljmarshall

**Comment by BewareMyPower** [comment]
It looks like it's because the JKS files are generated by a higher version JDK.

IMO, committing binaries files are terrible to maintain for developers. It's necessary to describe how to generate this files if we need to do so. Or should we add the steps to CI instead of uploading these binaries?

**Comment by nodece** [comment]
See https://stackoverflow.com/questions/1052700/java-io-ioexception-invalid-keystore-format

**Comment by Shoothzj** [comment]
@BewareMyPower what's your java8 version, I remember it needs higher java8 version.

**Comment by BewareMyPower** [comment]
1.8.0_261

**Comment by BewareMyPower** [comment]
> I remember it needs higher java8 version.

It's right. I switched to 1.8.0_332 and it works now.

**Comment by BewareMyPower** [comment]
> especially for things like certs that have expiration dates in the near future

Yeah, there was once a similar issue (https://github.com/apache/pulsar/pull/9607) that broke the whole CI.

**Comment by Shoothzj** [comment]
@michaeljmarshall @BewareMyPower Thanks for your advice, I will follow up on this.

**Comment by mattisonchao** [comment]
Hi @Shoothzj 
There are a lot of conflicts when this PR is cherry-picked to 2.9, Could you please help push a new PR to branch-2.9?

**Comment by Shoothzj** [comment]
@mattisonchao I am afraid I don't have time this week. But don't worry, @Lico-Tom will help open a new PR

**Comment by mattisonchao** [comment]
It looks like cherry-picked pull request a118ab953ede75d6ecfdf32aff95595f270486a1 already cover this change in branch-2.9. Could you have time to help check it?

**Comment by Shoothzj** [comment]
@mattisonchao Yes, It's already covered this change.

**Comment by BewareMyPower** [comment]
@mattisonchao @Shoothzj Does it mean there is no need to cherry-pick this PR? If yes, I think it would be better to remove all `release/xxx` labels and `cherry-picked/xxx` labels instead of marking this PR as "cherry-picked".

**Comment by BewareMyPower** [comment]
I removed these labels after confirming it from @mattisonchao.

**Pull request merged**